File: 2021년/8월/0805 최소비용 구하기 2.py
# 백준_1916_최소비용 구하기 2_다익스트라_골드 3

import sys
import heapq
input = sys.stdin.readline
INF = sys.maxsize

n = int(input())
m = int(input())
graph = [[]for _ in range(n+1)]
for _ in range(m):
    a, b, c = map(int, input().split())
    graph[a].append((b, c))
start, end = map(int, input().split())

dist = [INF]*(n+1)
dist[start] = 0
route = [0]*(n+1)

heap = []
heapq.heappush(heap, (0, start))
while heap:
    now_cost, now_urban = heapq.heappop(heap)

    if dist[now_urban] < now_cost:
        continue

    for next_urban, next_cost in graph[now_urban]:
        if dist[next_urban] > next_cost + now_cost:
            route[next_urban] = now_urban
            dist[next_urban] = next_cost + now_cost
            heapq.heappush(heap, (dist[next_urban], next_urban))

route_v = []
temp = end
while temp:
    route_v.append(temp)
    temp = route[temp]

print(dist[end])
print(len(route_v))
for i in range(len(route_v)-1, -1, -1):
    print(route_v[i], end=" ")

File: 2021년/8월/0805 최소비용 구하기 2.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <algorithm> 
using namespace std;

int INF = 1000000000;
int n, m, start, last;
int dist[1001];
int route[1001];
vector<pair<int, int> > graph[1001];
vector<int> route_v;
priority_queue<pair<int, int> > heap;

int main()
{
	scanf("%d %d",&n,&m);
	while(m--)
	{
		int a,b,c;
		scanf("%d %d %d",&a,&b,&c);
		graph[a].push_back(make_pair(b,c));
	}
	scanf("%d %d",&start,&last);
	
	fill(dist, dist+n+1, INF);
	dist[start] = 0;
	
	heap.push(make_pair(0, start));
	while(!heap.empty())
	{
		int now_cost = -heap.top().first;
		int now_urban = heap.top().second;
		heap.pop();
		
		if (dist[now_urban] < now_cost) continue;
		
		for(int i=0;i<graph[now_urban].size();i++)
		{
			int next_urban = graph[now_urban][i].first;
			int next_cost = graph[now_urban][i].second;
			if(dist[next_urban] > now_cost + next_cost)
			{
				route[next_urban] = now_urban;
				dist[next_urban] = now_cost + next_cost;
				heap.push(make_pair(-dist[next_urban], next_urban));
			} 
		}
	}
	
	int temp = last;
	while(temp)
	{
		route_v.push_back(temp);
		temp = route[temp];
	}
	
	printf("%d\n", dist[last]);
	printf("%d\n", route_v.size());
	for(int i=route_v.size()-1;i>=0;i--)
	{
		printf("%d ",route_v[i]);	
	}	
}
